kafka/consumer.go

```diff
@@ -329,15 +329,20 @@ func (c *Consumer) rebalance(ev Event) bool {
 // consumerReader reads messages and events from the librdkafka consumer queue
 // and posts them on the consumer channel.
 // Runs until termChan closes
-func consumerReader(c *Consumer, termChan chan bool) {
+func consumerReader(c *Consumer, termChan chan bool, eventsChanSize int) {
 
 out:
 	for true {
 		select {
 		case _ = <-termChan:
 			break out
 		default:
-			_, term := c.handle.eventPoll(c.events, 100, 1000, termChan)
+			_, term := c.handle.eventPoll(
+				c.events,
+				100,
+				eventsChanSize - len(c.events), // max events depends on available buffer
```

Comment:
Yeah, I agree on the channelSize > pollCnt case: we should try to fetch as many events as possible to fill the channel buffer, but I think we should be more careful with the reverse where the channel buffer size is smaller than the poll count.
For performance we want to avoid idle periods and I imagine that a small channelSize and pollCnt will result in idle periods when the application side is as fast as the Go client side:
 - Loop 1: Go client polls 100 entries from librdkafka and enqueues them on channel.
 - Loop 2: Go client calculates that pollCnt - len(chan)  == 0 and backs off with a delay
 - Application reads all events from the channel
 - Idle period while loop 2's backoff delay times out..
 - Go to loop 1

Compare this to the current behaviour:
 - Loop 1: Go client polls 100 entries from librdkafka and enqueues them on channel.
 - Loop 2: Go client pools another 100 entries from librdkafka but blocks on the first channel produce since the channel is full.
 - Application reads all events from the channel
 - Go client immediately unblocks and when channel space is available
 - Go to loop 1

There's probably a golden ratio where the pollCnt is some multiple of channelSize, but this depends on actual usage, so I feel the current approach with simply blocking provides the necessary and desired backpressure mechanism.
Having said that there is no point in polling 1000 events if the channelSize is 2, so some reasonable multiplier should be in place.
Maybe something like: `pollCnt = math.Max(channelSize * 3, 100)`